### DISCUSSION TITLE
4 settembre 2023 - esercizio 2

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#45] [Esami]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/b2156673-72e5-401d-99d7-aecb5ed079d4)

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/9ce1a068-b998-43e1-ab9d-7485dd0c58c8)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/c6746f87-9e33-4399-a0b1-27408a69584d)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/d8848bf0-550a-4458-b8ff-900b49eede3d)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/1414ffe4-9a6b-432b-aa94-74bf1f2ecd37)

**Reply by FrascaHoryzont** [reply]
quale criterio utilizzo per decidere le chiavi? perché pensavo che dovessi prendere gli elementi che non apparivano come determinante, ma nell'esercizio tutti gli elementi di R appaiono come determinante. Quindi non capisco con quale criterio vado a prendere AH, AB, BD.

**Reply by IonutCicio** [reply]
Si prendono gli elementi che non sono mai **determinati**, perché l'unico modo per determinarli in un'eventuale chiave è se sono nella chiave (non li determina nessun'altro se non se stessi per riflessività). Però non basta, questa è solo un'euristica, devi vedere la definizione di chiave.

Sia $R$ uno schema relazionale, e sia $F$ un insieme di dipendenze funzionali su $R$. Sia $K \subseteq R$, $K$ si dice chiave se
- $K \to R \in F^+$
- $\forall K' \subset K, K' \to R \notin F^+$

Un sottoinsieme di attributi $K$ si dice chiave se determina tutto lo schema $\iff R \subseteq (K)^+_F$ _(Lemma visto a lezione)._ Inoltre, se prendi $K' \subset K$, quindi $K'$ **sottoinsieme proprio** (vuol dire che non deve essere uguale a $K$), $K'$ non deve essere chiave. Vuol dire che $K$ è una chiave miniamle.

Prendiamo un esempio $(BD)^+_F = R \implies BD$ è superchiave _(determina tutti gli attributi dello schema)_. Ora dobbiamo verificare che $B$ e $D$ non siano chiavi. Infatti $(B)^+_F = B \neq R$ e $(D)^+_F = D \neq R$, quindi $BD$ è chiave.

**Reply by IonutCicio** [reply]
Oltretutto, questo esercizio è particolare perché non specifica quante sono le chiavi, potrebbe essere 1 _(si vede con il criterio dell'intersezione)_ o potrebbero essere 3, 4, 5 etc... infatti lui si è perso la chiave $AC$. Per vedersi tutte le chiavi, la soluzione più comoda è scriversi tutti i sottoinsiemi, ed eliminare quelli che non sono chiave man mano che si trovano chiavi / si calcolano le chiusure usando certe euristiche _(ad esempio, se ABCD non è chiave, neanche tutti i sottoinsiemi di ABCD sono chiave)_

**Reply by FrascaHoryzont** [reply]
Quindi per vederlo posso prendere l'insieme e considerare tutte le coppie possibili (AB,BC,CD, ecc...) e vedo tramite i loro sottoinsiemi (se già di per se sono chiavi) se sono chiavi oppure no giusto?

**Reply by IonutCicio** [reply]
> [!CAUTION]
> Nel messaggio originale avevo scritto $K' \to K \notin F^+$, è un errore, dovevo scrivere $K' \to R \notin F^+$

Comunque sì, devo verificare che i sottoinsiemi non sono chiavi. Peccato che considerare tutte le coppie possibili non basta. Una terna $ABE$ potrebbe essere benissimo chiave se $AB$, $BE$ e $AE$ non sono chiavi e $ABE$ determina $R$.

In questo esercizio $R = ABCDEH$. Sai che $E$ non può essere in una chiave perché non compare mai a sinistra, quindi ti rimangono da considerare le combinazioni degli attributi $ABCDH$, che sono $2^5 = 32$. Non devi verificare la chiusura per tutte, puoi usare qualche euristica per tagliarti un sacco di combinazioni una volta che hai calcolato una chiusura, ma è comunque laborioso se non sai di preciso quante sono le chiavi.

Se vuoi qui #24 ho una soluzione completa in cui calcolo tutte le chiavi di uno schema non sapendo quante sono in anticipo.

**Reply by FrascaHoryzont** [reply]
grazie mille vedo subito.

**Comment by niccolo-fato** [comment]
![IMAGE 2024-01-14 09:05:56](https://github.com/sapienzastudentsnetwork/bdi2324/assets/73995334/94811cd7-f3ff-40fd-9ab5-ef4392c2b039)
![IMAGE 2024-01-14 09:06:05](https://github.com/sapienzastudentsnetwork/bdi2324/assets/73995334/3c14bc03-59a4-4c64-836c-8900f2fe74a2)

**Reply by UnoCasuale** [reply]
Mi sembra corretto, ma credo tu ti sia perso la chiave AC

**Reply by niccolo-fato** [reply]
sì, hai ragione

**Comment by Frisbon** [comment]
Controllando i determinanti, vedo che E non determina nulla ⇒ non é / non fa parte di una chiave. Quindi le calcolo per ABCDH

Controllo le combinazioni:

$(A,B,C,D,H) ^+ \not= R$

Ora quelle a due…

$AB^+ = R ⇒ K$ ⇒ Non devo calcolare combinazioni con AB dentro.
$AC^+ = R ⇒ K$ ⇒ Same
$AD^+ ≠ R$
$AH^+ = R ⇒ K$ ⇒ Same
$BC^+ ≠ R$
$BD^+ = R ⇒ K$ ⇒ Same
$BH^+ ≠ R$
$CD^+ ≠ R$
$CH^+ ≠ R$
$DH^+ ≠  R$ 

Ora quelle a tre… (cancello quelle che contengono chiave)

~~ABC,ABD,ABH~~,~~ACD,ACH,ADH~~
~~BCD~~,BCH,~~BDH~~,CDH
$(BCH,CDH)^+ \not= R$

Ora quelle a quattro sono solo queste e non posso usarle ~~ABCD ACDH BCDH~~

Quindi le uniche chiavi sono:
**AB,AC,AH,BD**

---

3NF sse per ogni dip fun X→Y ho che X é chiave or Y in a chiave

C→E viola la 3NF. (dimostrazione lasciata per casa)

---

Cerco cop minima. Decompongo F
diventa *{B~~C~~D→A,BCD→H,BD→C,BD→H,A→D,C→E,CD→H,H→E,AH→B}* 

Ora controllo le ridondanze nei determinanti a coppia.
Se ho XYZ→A Calcolo XY,YZ,AX e vedo se (xes) A$\in$XY e se succede, levo Z. 

BCD→A ⇒ $BC^+ = BCE,\  BD^+ =R, \ CD^+ = DEH$, A apparteiene a BD quindi C é ridondante. Elimino
BCD→H ⇒ H appartiene a BD e CD quindi lascio (per ora)
BD→C ⇒ $B^+ = B,\ D^+= D$, C non appartiene a loro, lascio.
BD→H ⇒ Same
CD→H ⇒ $C^+=CE,\ D^+ = D$, H non appartiene a loro, lascio.
AH→B ⇒ $A^+ = AD,\ H^+=EH$, same, lascio.

Mi rimane:
*{BD→A,~~BCD→H~~,BD→C,~~BD→H~~,A→D,C→E,CD→H,H→E,AH→B}*

Ora, per ogni dipendenza, calcolo le chiusure senza quella stessa presa in considerazione.

$BD^+ = BCDEH$, A non ne fa parte, lascio…
$BCD^+ = R$, H ne fa parte, Elimino…
$BD^+ = BDHE$, C not in, lascio.
$BD^+= BDCEH$, H dentro, elimino…
$A^+ = A$, lascio
$C^+ = C$, same
$CD^+ = CDE$, H  not in, lascio.
$H^+= H$,same
$AH^+= ADHE$, b not in, lascio 

Quindi la copertura minima che rimane é…
*{BD→A,BD→C,A→D,C→E,CD→H,H→E,AH→B}*

Applico l’algoritmo di decomposizione…
Il primo for mi cerca attributi che non sono in F, ci sono tutti quindi $S=$ é vuoto ⇒ salto il primo if e diventa $\rho = \{\text{BDA, BDC, AD, CE, CDH, HE, AHB}\}$

**Comment by daromacarlo** [comment]
![Immagine 2024-02-11 215305](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/aac835d7-b386-47f3-9ba1-0a3414ee0039)
![Immagine 2024-02-11 215315](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/29c96730-c861-403c-a915-4a6c49fc0f3a)
![Immagine 2024-02-11 215326](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/368d4d58-925a-443e-96ea-1cb8b8a57fd9)
![Immagine 2024-02-11 215340](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/5dd5968d-8717-4e94-a9a7-c319b2140a51)
![Immagine 2024-02-11 215352](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/83837796-c948-4729-b72d-fc0895ce783e)
![Immagine 2024-02-11 215401](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/2bbf205e-e531-491d-9acf-8f3bff013bcd)
![Immagine 2024-02-11 215412](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/58d0abb0-2baa-4f8c-b2c6-d234b933436e)
![Immagine 2024-02-11 215421](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/8ccc69f4-2aed-418e-b007-d444774baf13)
![Immagine 2024-02-11 215429](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/d2a22c1a-714b-4774-85fd-407572fc53b6)

**Comment by sprigatita** [comment]
![image](https://github.com/user-attachments/assets/5c3755bb-d60f-4aa2-83ff-f48b5f3504e1)
![image](https://github.com/user-attachments/assets/f835de3f-c7d0-4f68-a992-530988b5a38b)
![image](https://github.com/user-attachments/assets/9399c267-02bb-4b08-8851-bf98aca13489)
